### PR TITLE
Try running Travis jobs in GitHub Actions

### LINKED ISSUE #8793: Start migrating off of Travis to GitHub Actions or CircleCI
In-depth discussion here: https://www.jeffgeerling.com/blog/2020/travis-cis-new-pricing-plan-threw-wrench-my-open-source-works

I'm going to try opening a PR using an alternative... i think the parallel jobs may be tough!

https://circleci.com/docs/2.0/migrating-from-travis/#section=getting-started - `Organizations on our free plan get 400,000 credits per month for open source builds.`

https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions/migrating-from-travis-ci-to-github-actions

### PR DESCRIPTION
**Pull request by jywarren**
starting with db setup, gems, yarn, and unit tests

Fixes #8793 

And lots more stuff related to chromedriver... maybe this is now added in?: 

https://github.com/publiclab/plots2/blob/ff1974a08b39aedc5982ac602e89f9a3c0f6c2d9/.travis.yml#L1-L39

### CODE REVIEW

**Comment by jywarren** [comment]
Now looking at https://docs.github.com/en/free-pro-team@latest/actions/guides/building-and-testing-ruby

**Comment by cesswairimu** [comment]
Hi @jywarren , is there any way I can help here?
Are we trying to make *continuous-integration* to pass or *travis CI* to finish  or both?

File: .github/workflows/continuous-integration.yml [hunk 1]
```diff
@@ -0,0 +1,108 @@
+name: continuous-integration
+on: [push]
+jobs:
```
Comment (jywarren) [inline .github/workflows/continuous-integration.yml:3]: Once we get this running we can look at relative runtimes of each test suite (unit, functional, integration, system) and try breaking them into parallel jobs to see if we can get the whole thing to run faster.

**Comment by jywarren** [comment]
Thanks @cesswairimu -- i'm trying to get all the tests that formerly ran in Travis to now run in Github Actions 😅 

We're almost there. Any ideas on the remaining failing tests? We are in MySQL now, not MariaDB, for one, but most of the rest should be identical!

We could also try running MariaDB again in here, if it helps... feel free to make commits or to make your own branch and PR!

**Comment by jywarren** [comment]
Oh, and sorry, for some reason i have been taking notes in https://github.com/publiclab/plots2/issues/8793 not here. I started and then i just kept doing it, sorry!

**Comment by jywarren** [comment]
OK - noting that we've seen all these in the past few commits, but oddly they have been spread out in the log whereas now they all occurred together; not sure if that's intermittent or what; they seem also to be a mix of integration and system test failures, of 3 types: 

for **integration tests**, 2 "wrong number of arguments, (given 3, expected 1)` and 4 action cable `no messages sent` errors

for **system tests**: 4 CSS capybara misses in the oddly-located `/test/editor_test.rb` system test.

Note that our attempt to 'add redis port to config/cable.yml' didn't seem to help in a recent commit, nor did a recent attempt to refine chromedriver config per https://github.com/marketplace/actions/setup-chromedriver?version=v1.0.1

```
ERROR["test_should_choose_i18n_for_tag/show", #<Minitest::Reporters::Suite:0x0000555dada05590 @name="I18nTest">, 245.096457783]
 test_should_choose_i18n_for_tag/show#I18nTest (245.10s)
Minitest::UnexpectedError:         ArgumentError: wrong number of arguments (given 3, expected 1)
            test/integration/I18n_test.rb:436:in `block (2 levels) in <class:I18nTest>'
            test/integration/I18n_test.rb:431:in `each'
            test/integration/I18n_test.rb:431:in `block in <class:I18nTest>'

=ERROR["test_should_choose_i18n_for_wiki/show", #<Minitest::Reporters::Suite:0x0000555db32eca50 @name="I18nTest">, 295.37204805500005]
 test_should_choose_i18n_for_wiki/show#I18nTest (295.37s)
Minitest::UnexpectedError:         ArgumentError: wrong number of arguments (given 3, expected 1)
            test/integration/I18n_test.rb:360:in `block (2 levels) in <class:I18nTest>'
            test/integration/I18n_test.rb:355:in `each'
            test/integration/I18n_test.rb:355:in `block in <class:I18nTest>'

=Capybara starting Puma...
* Version 5.0.4 , codename: Spoony Bard
* Min threads: 0, max threads: 4
* Listening on http://127.0.0.1:33475
[Screenshot]: tmp/screenshots/failures_test_check_markdown_editor_features_are_functional.png
ERROR["test_check_markdown_editor_features_are_functional", #<Minitest::Reporters::Suite:0x0000555db3105638 @name="EditorTest">, 307.38931888699994]
 test_check_markdown_editor_features_are_functional#EditorTest (307.39s)
Minitest::UnexpectedError:         Capybara::ElementNotFound: Unable to find css "a[data-original-title='Edit this wiki page.']"
            test/editor_test.rb:65:in `block in <class:EditorTest>'

[Screenshot]: tmp/screenshots/failures_test_check_rich_editor_features_are_functional.png
ERROR["test_check_rich_editor_features_are_functional", #<Minitest::Reporters::Suite:0x0000555db390d8a8 @name="EditorTest">, 311.46063991200003]
 test_check_rich_editor_features_are_functional#EditorTest (311.46s)
Minitest::UnexpectedError:         Capybara::ElementNotFound: Unable to find css "a[data-original-title='Try the beta inline Rich Wiki editor.'"
            test/editor_test.rb:41:in `block in <class:EditorTest>'

[Screenshot]: tmp/screenshots/failures_test_check_that_markdown_wiki_editor_functions_correctly.png
ERROR["test_check_that_markdown_wiki_editor_functions_correctly", #<Minitest::Reporters::Suite:0x0000555db3837190 @name="EditorTest">, 315.5769444949999]
 test_check_that_markdown_wiki_editor_functions_correctly#EditorTest (315.58s)
Minitest::UnexpectedError:         Capybara::ElementNotFound: Unable to find css "a[data-original-title='Edit this wiki page.'"
            test/editor_test.rb:31:in `block in <class:EditorTest>'

[Screenshot]: tmp/screenshots/failures_test_check_that_rich_wiki_editor_functions_correctly.png
ERROR["test_check_that_rich_wiki_editor_functions_correctly", #<Minitest::Reporters::Suite:0x0000555db378df00 @name="EditorTest">, 319.7087934880001]
 test_check_that_rich_wiki_editor_functions_correctly#EditorTest (319.71s)
Minitest::UnexpectedError:         Capybara::ElementNotFound: Unable to find css "a[data-original-title='Try the beta inline Rich Wiki editor.'"
            test/editor_test.rb:17:in `block in <class:EditorTest>'

===== FAIL["test_broadcast_only_to_users_with_notification_all_tag", #<Minitest::Reporters::Suite:0x0000555db509ec10 @name="CommentNotificationTest">, 321.8890391130001]
 test_broadcast_only_to_users_with_notification_all_tag#CommentNotificationTest (321.89s)
        No messages sent with {:notification=>{:title=>"New Comment on This is purple-air with hyphen", :path=>"/notes/admin/02-20-2019/purple-air-with-hyphen", :option=>{:body=>"Hey, @naman18996, @jeffrey and @jeff. This is Action Cable Integration Test", :icon=>"https://publiclab.org/logo.png"}}} to users:notification:13
        test/action_cable/integration/comment_notification_test.rb:203:in `test_broadcast_only_to_users_with_notification_all_tag'

= FAIL["test_broadcasts_when_user_is_loggedin", #<Minitest::Reporters::Suite:0x0000555db54d4f00 @name="CommentNotificationTest">, 322.02027429500004]
 test_broadcasts_when_user_is_loggedin#CommentNotificationTest (322.02s)
        No messages sent with {:notification=>{:title=>"New Comment on This is purple-air with hyphen", :path=>"/notes/admin/02-20-2019/purple-air-with-hyphen", :option=>{:body=>"Hey, @naman18996 and @jeffrey. This is Action Cable Integration Test", :icon=>"https://publiclab.org/logo.png"}}} to users:notification:13
        test/action_cable/integration/comment_notification_test.rb:98:in `test_broadcasts_when_user_is_loggedin'

 FAIL["test_no_broadcast_to_non_engaged_user", #<Minitest::Reporters::Suite:0x0000555db5976b30 @name="CommentNotificationTest">, 322.1566322230001]
 test_no_broadcast_to_non_engaged_user#CommentNotificationTest (322.16s)
        No messages sent with {:notification=>{:title=>"New Comment on This is purple-air with hyphen", :path=>"/notes/admin/02-20-2019/purple-air-with-hyphen", :option=>{:body=>"Hey, @naman18996 and @jeffrey. This is Action Cable Integration Test", :icon=>"https://publiclab.org/logo.png"}}} to users:notification:13
        test/action_cable/integration/comment_notification_test.rb:156:in `test_no_broadcast_to_non_engaged_user'

 FAIL["test_number_of_broadcasts_when_user_is_loggedin", #<Minitest::Reporters::Suite:0x0000555db6026688 @name="CommentNotificationTest">, 322.35120182900005]
 test_number_of_broadcasts_when_user_is_loggedin#CommentNotificationTest (322.35s)
        No messages sent with {:notification=>{:title=>"New Comment on This is purple-air with hyphen", :path=>"/notes/admin/02-20-2019/purple-air-with-hyphen", :option=>{:body=>"Hey, @naman18996 and @jeffrey. This is Action Cable Integration Test", :icon=>"https://publiclab.org/logo.png"}}} to users:notification:13
        test/action_cable/integration/comment_notification_test.rb:128:in `test_number_of_broadcasts_when_user_is_loggedin'

==============|
```

**Comment by jywarren** [comment]
Noting via @noi5e @RuthNjeri this happening in current `main` branch when run locally; presumably these weren't an issue in Travis but if any match we can potentially address them or even defer them if we have to, to unblock:

```
2 failures in test/unit/api/search_service_test.rb (lines 56 & 71)
1 failure in test/unit/node_shared_test.rb (line 24)
1 failure in test/functional/search_controller_test.rb (line 51)
```

Hmm. It doesn't look like there are any matches. Too bad!!!

**Comment by jywarren** [comment]
OK, taking on the `ArgumentError: wrong number of arguments (given 3, expected 1)` errors; it is the same error as https://github.com/publiclab/plots2/pull/5948#issuecomment-508191391, https://github.com/publiclab/plots2/pull/5948#issuecomment-508560604, where @CleverFool77 was able to bypass it by using a custom attribute just for this test. 

But the situation while similar is not identical. Here are our two, on line 360 and 435:

https://github.com/publiclab/plots2/blob/80648fea3af7075d4834d22a837fa5d5f0aa5ec3/test/integration/I18n_test.rb#L354-L362

https://github.com/publiclab/plots2/blob/80648fea3af7075d4834d22a837fa5d5f0aa5ec3/test/integration/I18n_test.rb#L430-L438

Those two are the only ones with nested CSS selectors. I've tried inserting `>` to make the nesting more explicit. But if it doesn't work we can make the CSS more specific with additional classes. 451578e

**Comment by jywarren** [comment]
I'll also try moving that weird `editor_test.rb` into the system tests folder...

**Comment by jywarren** [comment]
Next up; solve the ActionCable test problem?

**Comment by jywarren** [comment]
I think we may need to confirm that the Action Cable tests are running as part of the integration testing step.

Also, maybe `REDIS_HOST: localhost` instead, depending on the setup

Finally, we should check what env vars we are relying on in our ActionCable tests! the examples given by GitHub docs are for REDIS_HOST and REDIS_PORT but we may use different ones!

File: test/integration/I18n_test.rb [hunk 2]
```diff
@@ -433,7 +433,7 @@ class I18nTest < ActionDispatch::IntegrationTest
       follow_redirect!
 
       get '/tag/some-tag'
-      assert_select 'a span.d-none', I18n.t('tag.show.wiki_pages')
+      assert_select 'a > span.d-none', I18n.t('tag.show.wiki_pages')
```
Comment (jywarren) [inline test/integration/I18n_test.rb:436]: This is the last remaining assert_select problem, it's now a failure, not an error!

```
==== FAIL["test_should_choose_i18n_for_tag/show", #<Minitest::Reporters::Suite:0x000055fc54390868 @name="I18nTest">, 122.09785266000006]
 test_should_choose_i18n_for_tag/show#I18nTest (122.10s)
        Expected at least 1 element matching "a > span.d-none", found 0..
        Expected 0 to be >= 1.
        test/integration/I18n_test.rb:436:in `block (2 levels) in <class:I18nTest>'
        test/integration/I18n_test.rb:431:in `each'
        test/integration/I18n_test.rb:431:in `block in <class:I18nTest>'
```

we need to add `d-lg-inline` in between those two nests

**Comment by jywarren** [comment]
OK, noting we got that intermittent test failure at: 

```
Progress: |============ FAIL["test_that_pinned_notes_with_\"pinned:foo\"_tags_appear_at_the_top_of_[nodes:foo]_inline_tables", #<Minitest::Reporters::Suite:0x0000556cf9d46d98 @name="NodeSharedTest">, 2.527351892000013]
 test_that_pinned_notes_with_"pinned:foo"_tags_appear_at_the_top_of_[nodes:foo]_inline_tables#NodeSharedTest (2.53s)
        Expected: 1
          Actual: 0
        test/unit/node_shared_test.rb:24:in `block in <class:NodeSharedTest>'
```

☹️ Restarted it tho!

File: test/integration/I18n_test.rb (hunk 2 above)
Comment (jywarren) [reply test/integration/I18n_test.rb]: Hmmm. Looks like these actually passed while during integration test phase, but then somehow get run again during system tests, and then fail then??? Strange!

```

ERROR["test_should_choose_i18n_for_tag/show", #<Minitest::Reporters::Suite:0x000056155d689b60 @name="I18nTest">, 155.05812749199993]
 test_should_choose_i18n_for_tag/show#I18nTest (155.06s)
Minitest::UnexpectedError:         ArgumentError: wrong number of arguments (given 3, expected 1)
            test/integration/I18n_test.rb:436:in `block (2 levels) in <class:I18nTest>'
            test/integration/I18n_test.rb:431:in `each'
            test/integration/I18n_test.rb:431:in `block in <class:I18nTest>'

=ERROR["test_should_choose_i18n_for_wiki/show", #<Minitest::Reporters::Suite:0x000056155159ba48 @name="I18nTest">, 207.88893608799992]
 test_should_choose_i18n_for_wiki/show#I18nTest (207.89s)
Minitest::UnexpectedError:         ArgumentError: wrong number of arguments (given 3, expected 1)
            test/integration/I18n_test.rb:360:in `block (2 levels) in <class:I18nTest>'
            test/integration/I18n_test.rb:355:in `each'
            test/integration/I18n_test.rb:355:in `block in <class:I18nTest>'
```

File: .github/workflows/continuous-integration.yml [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        DB_PASSWORD: root
+        DB_PORT: ${{ job.services.mysql.ports[3306] }}
+      run: bundle exec rails test test/unit
+
+    - name: "Functional Tests"
+      env:
+        RAILS_ENV: test
+        DB_PASSWORD: root
+        DB_PORT: ${{ job.services.mysql.ports[3306] }}
+        REDIS_HOST: redis
```
Comment (icarito) [inline .github/workflows/continuous-integration.yml:93]: I think `localhost` should be right, but note that this variable is also defined here.

File: test/integration/I18n_test.rb (hunk 2 above)
Comment (jywarren) [reply test/integration/I18n_test.rb]: aha i had the system test command syntax wrong, fixed it. it had been running all tests twice, i think, on the system test step!

File: .github/workflows/continuous-integration.yml (hunk 3 above)
Comment (jywarren) [reply .github/workflows/continuous-integration.yml:93]: aha ok great, thanks! I'm not sure Redis is used in this suite so if I don't see errors again i may just remove these here... let's see!

Comment (jywarren) [reply .github/workflows/continuous-integration.yml:93]: ah yes, it's not erroring, so i'll try removing them here and relying on the system test ones only!

**Comment by jywarren** [comment]
OK all tests are passing except for:

```

[Screenshot]: tmp/screenshots/failures_test_check_markdown_editor_features_are_functional.png
ERROR["test_check_markdown_editor_features_are_functional", #<Minitest::Reporters::Suite:0x00007f34518414d0 @name="EditorTest">, 90.52515762799999]
 test_check_markdown_editor_features_are_functional#EditorTest (90.53s)
Minitest::UnexpectedError:         Capybara::ElementNotFound: Unable to find css "a[data-original-title='Edit this wiki page.']"
            test/system/editor_test.rb:66:in `block in <class:EditorTest>'

[Screenshot]: tmp/screenshots/failures_test_check_rich_editor_features_are_functional.png
ERROR["test_check_rich_editor_features_are_functional", #<Minitest::Reporters::Suite:0x00007f345b370b40 @name="EditorTest">, 153.1470894050001]
 test_check_rich_editor_features_are_functional#EditorTest (153.15s)
Minitest::UnexpectedError:         Capybara::ElementNotFound: Unable to find css "a[data-original-title='Try the beta inline Rich Wiki editor.'"
            test/system/editor_test.rb:42:in `block in <class:EditorTest>'

[Screenshot]: tmp/screenshots/failures_test_check_that_markdown_wiki_editor_functions_correctly.png
ERROR["test_check_that_markdown_wiki_editor_functions_correctly", #<Minitest::Reporters::Suite:0x00007f345ad811a8 @name="EditorTest">, 215.64896502800002]
 test_check_that_markdown_wiki_editor_functions_correctly#EditorTest (215.65s)
Minitest::UnexpectedError:         Capybara::ElementNotFound: Unable to find css "a[data-original-title='Edit this wiki page.'"
            test/system/editor_test.rb:32:in `block in <class:EditorTest>'

=[Screenshot]: tmp/screenshots/failures_test_check_that_rich_wiki_editor_functions_correctly.png
ERROR["test_check_that_rich_wiki_editor_functions_correctly", #<Minitest::Reporters::Suite:0x00007f345b28f118 @name="EditorTest">, 278.21773055100016]
 test_check_that_rich_wiki_editor_functions_correctly#EditorTest (278.22s)
Minitest::UnexpectedError:         Capybara::ElementNotFound: Unable to find css "a[data-original-title='Try the beta inline Rich Wiki editor.'"
            test/system/editor_test.rb:18:in `block in <class:EditorTest>'

========[Screenshot]: tmp/screenshots/test_blog.png
```

**Comment by jywarren** [comment]
Made tweaks to the system test login sequence and can continue with more config changes to match post_test.rb...

**Comment by jywarren** [comment]
Oops, codecov upload failure. restarted.

File: test/unit/node_shared_test.rb [hunk 4]
```diff
@@ -15,6 +15,8 @@ class NodeSharedTest < ActiveSupport::TestCase
     before = "Here are some nodes in a table: \n\n[nodes:test] \n\nThis is how you make it work:\n\n`[nodes:tagname]`\n\n `[nodes:tagname]`\n\nMake sense?"
     nodes(:one).add_tag('pinned:test', User.first)
     nodes(:one).add_tag('test', User.first) # ensure it would appear anyways (although we aren't yet asserting order below, we should)
+    assert nodes(:one).has_tag('pinned:test')
+    assert nodes(:one).has_tag('test')
```
Comment (jywarren) [inline test/unit/node_shared_test.rb:19]: These passed but the lower one on 26 didn't... ☹️

**Comment by jywarren** [comment]
No luck fixing those `editor_test.rb` errors... Hmm.

**Comment by jywarren** [comment]
Hmm -- `<p><a class='btn btn-sm btn-outline-secondary' rel='tooltip' title='Try the beta inline Rich Wiki editor.' data-placement='top' href='/wiki/developers?raw=true'>` isn't fulfilling this, maybe?

https://github.com/publiclab/plots2/blob/edcac15264ab1c575b368e354131a426db7d42b2/test/system/editor_test.rb#L17

**Comment by jywarren** [comment]
I'm going to turn off the failing tests and re-open them in a follow-up.

**Comment by cesswairimu** [comment]
Yeeih 🎉🎉🎉🎉🎉 thanks so much Jeff

**Pull request merged**